[WIP] Rename char_list to charlist

eksperimental opened the pull request:
This is a big change.
We need to make consistent char_list by renaming it to charlist  

### DONE
- [x] Replace every occurrence of "char_list" and "char list"
- [x] Soft-deprecate `Kernel.to_char_list/1` and in `to_char_list` in /lib/elixir/lib/list/chars.ex
__Soft-deprecate:__
- [x] `/lib/elixir/lib/inspect/algebra.ex`: Inspect.Opts: :char_lists, :as_char_lists options
- [x] `/lib/elixir/pages/Typespecs.md`: Should we leave a note saying `char_list()` type has been deprecated in favor of `charlist()`?
- [x] `/lib/elixir/src/elixir.erl`, `/lib/elixir/src/elixir_rewrite.erl`: We need to softdeprecate `char_list/0` type
- [x] Fix protocol warnings when compiling Elixir

### TODO
- ~~~Add tests for soft-deprecated:~~~
  - ~~~functions / protocols: Kernel.to_char_list/1, Lists.Chars.to_char_list/1, File.open/1-2, and Inspect.~~~
  - ~~~options in algebra function: Inspect.Opts~~~
  - ~~~data type `char_list/0`~~~

josevalim commented:
@eksperimental this branch unfortunately has the commits from the unicode branch. I would recommend you to create a new branch from Elixir master and cherry pick the char list changes:

    git checkout elixir-lang/master
    git pull --rebase
    git checkout -b ek-charlists
    git cherry-pick 594aefa
    git cherry-pick d904e0a

This being git, there are probably other ways of doing the same. :) Let me know once you "rebase" so I can review this more throughly, great work so far.

josevalim commented:
Before hand, a couple notes:

1. we are not going to emit any warning so far, but we should add `@doc false` to the old versions (like `String.to_char_list`) and leave a note on top: `# TODO: Deprecate in v1.5`.

eksperimental commented:
@josevalim there we go. Please let me know what you think

eksperimental commented on lib/elixir/lib/inspect/algebra.ex:
```diff
@@ -44,19 +44,23 @@ defmodule Inspect.Opts do
                binaries: :infer | :as_binaries | :as_strings,
-               char_lists: :infer | :as_lists | :as_char_lists,
+               charlists: :infer | :as_lists | :as_charlists | :as_char_lists,
+               char_lists: :infer | :as_lists | :as_charlists | :as_char_lists,
```
I don't know if this is the right way, or if there is a way to hide :char_lists from the user documentation.
On a side note, should we do instead:
```elixir
charlists: :infer | :as_lists | :as_charlists,
char_lists: :infer | :as_lists | :as_charlists,
```
So if user switch to :charlists, we force him to use :as_charlists

josevalim replied:
I agree. :)

eksperimental commented on lib/elixir/lib/inspect/algebra.ex:
```diff
@@ -44,19 +44,23 @@ defmodule Inspect.Opts do
   defstruct structs: true,
             binaries: :infer,
+            charlists: :infer,
             char_lists: :infer,
```
Wondering if there is a way to hide this from user docs

josevalim replied:
Nothing we can, it is OK to keep it as is though. :)

josevalim commented on lib/elixir/lib/list/chars.ex:
```diff
@@ -3,55 +3,64 @@ defprotocol List.Chars do
+
+  # TODO: Deprecate in v1.5
+  @doc false
   def to_char_list(thing)
```
This will unfortunately require developers to implement both functions. You can try this:

```elixir
    Kernel.def to_char_list(thing) do
      to_charlist(thing)
    end
```

eksperimental replied:
May I ask why do we need to use `Kernel.def` and not just plain `def`?

eksperimental replied:
self answer:
== Compilation error on file lib/list/chars.ex ==
** (CompileError) lib/list/chars.ex:16: undefined function def/2

josevalim replied:
We are basically cheating. protocols remove `def/2` because we are not supposed to define functions on them (but it is pretty cool we can still do it :D).

eksperimental replied:
The language never stops surprising me :D

josevalim commented on lib/elixir/lib/list/chars.ex:
```diff
@@ -3,55 +3,64 @@ defprotocol List.Chars do
   end
 end
+
+# TODO: Deprecate in v1.5
```
If we do what I mentioned above, we can get rid of this line.

josevalim commented on lib/elixir/pages/Typespecs.md:
```diff
@@ -83,7 +83,7 @@ Built-in type           | Defined as
 `char()`                | `0..0x10ffff`
 `number()`              | `integer()` \| `float()`
-`char_list()`           | `[char()]`
+`charlist()`           | `[char()]`
```
Indentation. :)

josevalim commented on lib/elixir/src/elixir_rewrite.erl:
```diff
@@ -182,9 +189,11 @@ rewrite(?access, _DotMeta, 'get', Meta, [Arg, _], Env)
+    when is_list(List) andalso (Right == 'to_charlist' orelse Right == 'to_char_list') ->
   List;
-rewrite(?string_chars, _DotMeta, 'to_string', _Meta, [String], _Env) when is_binary(String) ->
+  rewrite(?string_chars, _DotMeta, 'to_string', _Meta, [String], _Env) when is_binary(String) ->
```
Indentation.

josevalim commented on lib/elixir/src/elixir_rewrite.erl:
```diff
@@ -182,9 +189,11 @@ rewrite(?access, _DotMeta, 'get', Meta, [Arg, _], Env)
     ['Elixir.Macro':to_string(Arg)]);
-rewrite(?list_chars, _DotMeta, 'to_char_list', _Meta, [List], _Env) when is_list(List) ->
+%% TODO: Deprecate in v1.5 - to_char_list function
+rewrite(?list_chars, _DotMeta, Right, _Meta, [List], _Env)
```
Let's define two clauses instead of defining complex guards. :)

josevalim commented on lib/elixir/lib/file.ex:
```diff
@@ -1318,7 +1318,7 @@ defmodule File do
   @read_ahead 64*1024
 
-  defp open_defaults([:char_list | t], _add_binary) do
+  defp open_defaults([:charlist | t], _add_binary) do
```
We need to add a clause that also handles `:char_list` and add a TODO. :)

josevalim commented:
This is excellent! I have added some tiny comments and we should be good to go!

eksperimental commented:
@josevalim all suggestions addressed, please have a look

eksperimental commented on lib/logger/lib/logger/utils.ex:
```diff
@@ -153,7 +153,7 @@ defmodule Logger.Utils do
 
   defp collect_cc(:encoding, [?l | t], args, used_format, used_args, opts),
-    do: collect_cc(:done, t, args, [?l | used_format], used_args, %{opts | char_lists: false})
+    do: collect_cc(:done, t, args, [?l | used_format], used_args, %{opts | charlists: false})
```
I don't understand what this is trying to do, what function does this opts get passed to? If it's inspect, then `false` is not a valid option, maybe it's `:as_lists` is what needs to be used..

josevalim replied:
You are right, `:as_lists` should be it. I think `false` works by accident, a test should actually tell us if `:as_lists` is correct or not.

eksperimental replied:
ok. I will replace :false wil :as_lists and submit a different PR to check the validity of Inspect.Opts

eksperimental replied:
actually I will correct the value in the upcoming PR.. so i don't mess things around

eksperimental commented on lib/elixir/lib/file.ex:
```diff
@@ -946,8 +946,8 @@ defmodule File do
 
-    * `:char_list` - when this term is given, read operations on the file will
-      return char lists rather than binaries.
+    * `:charlist` - when this term is given, read operations on the file will
```
Just for another PR: :charlist is not an option in `mode/0` type, so we need to correct that.

josevalim replied:
Yes, please.

eksperimental commented:
Where should we place the tests for the deprecated functions..?
should it contain a file of its own, or should it be along the existing tests?

josevalim commented:
I have added the last replies. :)

eksperimental commented:
great @josevalim  please see my question above yours.. about tests... I would like to do this as soon as possible since this change involves so many files, I'm afraid other PRs will conflict with this one, and I would like to have this out of the way ASAP

lexmag commented:
I believe we don't test deprecated functions especially if they are simply delegated calls.

josevalim commented:
Yes, no tests there. I think this looks ready to merge, right?

lexmag commented on lib/elixir/lib/code.ex:
```diff
@@ -594,10 +594,10 @@ defmodule Code do
 
       # Get the documentation for the first function listed
-      iex> [fun | _] = Code.get_docs(Atom, :docs) |> Enum.sort()
+      iex> [fun | _] = Code.get_docs(Atom, :docs) |> Enum.sort() |> tl
```
Why do we need `tl` here?

lexmag replied:
We are saying "Get the documentation for the first function listed".
We should rather change expected result in doctest.

eksperimental replied:
with these changes the first function is Atoms.to_char_list/1, that's why..

lexmag replied:
I think we need to choose other module name then.

eksperimental replied:
let's change it to get :moduledoc

lexmag commented on lib/elixir/src/elixir.erl:
```diff
@@ -10,7 +10,9 @@
 
 %% Top level types
--export_type([char_list/0, struct/0, as_boolean/1, keyword/0, keyword/1]).
+%% TODO: Deprecate in v1.5 - type char_list
```
Maybe `TODO: Deprecate char_list type in v1.5` reads better?

eksperimental replied:
it does read much better. I just didn't want to break the "TODO: Deprecate in v1.5" so a simple search would list all changes for v1.5.
But if we agree on the change I can do it,

lexmag replied:
My search patter for deprecations is `TODO: Deprecate` regardless of version and filter results for specific version later.

eksperimental replied:
sure,, I think we can change that.

lexmag commented on lib/elixir/lib/kernel/typespec.ex:
```diff
@@ -877,13 +881,14 @@ defmodule Kernel.Typespec do
 
-  defp typespec({:char_list, _meta, []}, vars, caller) do
-    typespec((quote do: :elixir.char_list()), vars, caller)
+  # TODO: Deprecate by v1.5 - type char_list
```
Maybe `TODO: Deprecate char_list type by v1.5` reads better?

lexmag commented on lib/elixir/lib/kernel/typespec.ex:
```diff
@@ -634,8 +636,10 @@ defmodule Kernel.Typespec do
   # Special shortcut(s)
-  defp typespec_to_ast({:remote_type, line, [{:atom, _, :elixir}, {:atom, _, :char_list}, []]}) do
-    typespec_to_ast({:type, line, :char_list, []})
+  # TODO: Deprecase in v1.5 - type "char_list"
```
Same as https://github.com/elixir-lang/elixir/pull/4582/files#r62434607.

lexmag commented on lib/elixir/lib/inspect/algebra.ex:
```diff
@@ -44,18 +44,22 @@ defmodule Inspect.Opts do
             pretty: false,
             safe: true
 
+  # TODO: Deprecase in v1.5 option ":as_char_lists" and "char_lists" key.
```
Maybe `Deprecate <what> <when>` reads better? :bowtie:

lexmag commented on lib/elixir/lib/inspect/algebra.ex:
```diff
@@ -44,18 +44,22 @@ defmodule Inspect.Opts do
 
   """
 
+  # TODO: Deprecase in v1.5 option "char_lists" key
```
Same as https://github.com/elixir-lang/elixir/pull/4582/files#r62434660.

lexmag commented on lib/elixir/lib/inspect.ex:
```diff
@@ -279,9 +279,22 @@ end
   def inspect([], _opts), do: "[]"
 
-  def inspect(thing, %Inspect.Opts{char_lists: lists} = opts) do
+  # TODO: Deprecase in v1.5 :char_lists, :as_char_lists
```
Same as https://github.com/elixir-lang/elixir/pull/4582/files#r62434660.

lexmag commented on lib/elixir/lib/inspect.ex:
```diff
@@ -279,9 +279,22 @@ end
-  def inspect(thing, %Inspect.Opts{char_lists: lists} = opts) do
+  # TODO: Deprecase in v1.5 :char_lists, :as_char_lists
+  def inspect(thing, %Inspect.Opts{charlists: lists, char_lists: lists_deprecated} = opts) do
+    lists = cond do
```
This `cond` can be replaced with `if`.

eksperimental replied:
absolutely, and we don't need an else clause

eksperimental replied:
actually, we do need the else clause.
FIXED

eksperimental commented:
@lexmag I have addressed all your suggestions,
If you are interested in having a quick overview of the "TODO: Deprecated"  lines, please have a look here https://gist.github.com/eksperimental/0066cbeed2c9f400e8d5fe999119da18

eksperimental commented:
If no more suggestions, the commit is ready to by merged

lexmag commented:
Looks great. 💛

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

eksperimental commented:
great! thank you @josevalim and @lexmag for your inputs
